src/exchange.ts

```diff
@@ -0,0 +1,235 @@
+   */
+  async addPair(pair: TokenPair, tags: DecodedTag[] = []): Promise<string> {
+    const contract = this.smartweave
+      .contract(this.utils.CLOB_CONTRACT)
+      .connect(this.wallet);
+
+    if (pair.length !== 2) throw new Error("Invalid pair. Length should be 2.");
+
+    pair.forEach((hash) => {
+      if (!this.utils.validateHash(hash))
+        throw new Error(`Invalid token address in pair "${hash}".`);
+    });
+
+    const interactionID = await cacheContractHook(
+      () =>
+        contract.writeInteraction(
+          {
+            function: "addPair",
+            pair,
+          },
+          [
+            {
+              name: "Exchange",
+              value: "Verto",
+            },
+            {
+              name: "Action",
+              value: "AddPair",
+            },
+            ...tags,
+          ]
+        ),
+      this.utils.CLOB_CONTRACT
+    );
+
+    if (!interactionID) throw new Error("Could not add pair.");
+
+    return interactionID;
+  }
+
+  /**
+   * Create a swap
+   * @param pair The two tokens to trade between. Must be an existing pair
+   * @param amount The amount of tokens sent to the contract
+   * @param price Optional price for the order
+   * @param tags Optional custom tags
+   * @returns OrderID
+   */
+  async swap(
+    pair: SwapPairInterface,
+    amount: number,
+    price?: number,
+    tags: DecodedTag[] = []
+  ): Promise<string> {
+    // Validate hashes
+    if (
+      !this.utils.validateHash(pair.from) ||
+      !this.utils.validateHash(pair.to)
+    )
+      throw new Error(
+        "Invalid ID in pair. Must be a valid SmartWeave contract ID"
+      );
+
+    const contract = this.smartweave
+      .contract(this.utils.CLOB_CONTRACT)
+      .connect(this.wallet);
+
+    const orderID = await cacheContractHook(async () => {
+      // Transfer input tokens to the orderbook
+      const transferID = await this.token.transfer(
+        amount,
+        pair.from,
+        this.utils.CLOB_CONTRACT,
+        [{ name: "Type", value: "Send-Input" }]
+      );
+
+      // Create the swap interaction
+      return await contract.writeInteraction(
+        {
+          function: "createOrder",
+          transaction: transferID,
+          pair: [pair.from, pair.to],
+          price: price,
+        },
+        [
+          {
+            name: "Exchange",
+            value: "Verto",
+          },
+          {
+            name: "Action",
+            value: "Order",
+          },
+          ...tags,
+        ]
+      );
+    }, [pair.from, this.utils.CLOB_CONTRACT, pair.to]);
+
+    if (orderID === null) throw new Error("Could not create order");
+
+    // Create exchange fee
+    await this.utils.createFee(amount, pair.from, orderID, "exchange");
+
+    // Create VRT holder fee
+    await this.utils.createFee(amount, pair.from, orderID, "token_holder");
+
+    // Call Discord hook
+    axios.post(`https://hook.verto.exchange/api/transaction?id=${orderID}`);
```

Comment:
Isolate this in a utility function. Also, the base for this URL should be in a constant, something like

```
class Constants { 
  public static VERTO_HOOK_URL: string = "https://hook.verto.exchange/";
}
```

Reply:
We are overcomplicating things. I don't think we need to isolate these, it is just a oneliner

Reply:
Is this going to be used in more parts of this system ? If you see a possibility that it will be used in other parts of this codebase, then it's not overcomplicating. But If you are sure, then it'll be alright.

Reply:
Yep, but that won't be used in other parts. This is actually something that we implemented a year ago, no updates since. The hook is only for new orders.